### ISSUE TITLE
getChangesForBuild listing wrong changes when builds are not in same order as submission to VCS

### ISSUE THREAD

**Issue by kaloth** [buildbot/buildbot#5060]
Hi folks! I'm really loving buildbot but there's one bug that really impacts it's usefulness in our environment.

When listing 'Contributers' to a build the wrong list of names is returned if builds are not performed chronologically (ie: in the same order as they were submitted to version control).

After some investigation I can see that the root cause is the "ChangesConnectorComponent.getChangesForBuild" function. It does this..

```
 # Get the last successful build on the same builder
        previousBuild = yield self.master.db.builds.getPrevSuccessfulBuild(currentBuild['builderid'],
                                                                           currentBuild[
                                                                               'number'],
                                                                           ssBuild)
        if previousBuild:
            for ss in (yield gssfb(previousBuild['id'])):
                toChanges[ss['codebase']] = yield self.getChangeFromSSid(ss['ssid'])
```

Which just gets the previous thing successfully built. It then goes on to list the changes between this build and the current one.

In a simple setup this is ok but in one where builds can be out of order it causes chaos in the blame lists (and a lot of angry emails in reply!).

It may be better to instead iterate the changes backwards starting from the current build's change until you find a change with an associated passing build. That would then always give you the correct list of changes.

Any thoughts?

**Comment by tardyp** [comment]
Indeed, this is a very good comment. I didn't think about this way.
I am not sure how you would have the build not process more or less in the same order of the build, but anyway we could try to make a sql query which looks at the list of changes of the same branch and the same repository.

For multi codebase systems this can be a little bit tricky though you have to go back in time progressively in the whole code base until you find a build (actually buildset) associated with them which is successful.

Don't hesistate to have a try on it. Buildbot can only improve with its community trying to fix their own real life issue.

**Comment by kaloth** [comment]
Thanks tardyp! I'll have a go at fixing it and you can tell me if I've done it properly 😄
